upload-api/external-services/sso-providers/index.js

```
@@ -0,0 +1,179 @@
+            issuer: this.serviceSigner,
+            ran: input.invocation,
+            result: ok({
+              expiration: Math.floor(Date.now() / 1000) + lifetimeInSeconds,
+              request: input.invocation.cid
+            })
+          })
+        ]
+      })
+      const messageWriteRes = await this.agentStore.messages.write({
+        source: await Transport.outbound.encode(message),
+        data: message,
+        index: AgentMessage.index(message)
+      })
+
+      if (messageWriteRes.error) {
+        console.error(messageWriteRes.error)
+        return error(new Error('failed to write access/authorize invocation and receipt for SSO provider ' + authProvider))
+      }
+      const customer = DidMailto.fromEmail(/** @type {`${string}@${string}`} */(result.ok.userData.email))
+          
+      const confirmRes = await Access.confirm
+        .invoke({
+          issuer: this.serviceSigner,
+          audience: this.serviceSigner,
+          with: this.serviceSigner.did(),
+          lifetimeInSeconds,
+          nb: {
+            iss: customer,
+            att: input.capability.nb.att,
+            aud: input.capability.with,
+            // Link to the invocation that requested the SSO authorization.
+            cause: input.invocation.cid,
+          },
+        })
+        .execute(uploadServiceConnection())
+
+      if (confirmRes.out.error) {
+        console.error('executing access/confirm', confirmRes.out.error)
+        return error(new Error('failed to execute access/confirm invocation for SSO provider ' + authProvider))
+      }
+
+      // Note: here we can add the customer to the customer store here with a trial plan to eliminate the need of the plan selection step
+      
+      // Step 3: Return the link to the confirmed authorization
+      return ok(confirmRes.link())
+    } catch (err) {
+      return error(new Error(`SSO provider ${authProvider} failed: ${err instanceof Error ? err.message : String(err)}`))
+    }
```

Let this throw - there's nothing in the `try` that doesn't return a result.